### ISSUE TITLE
Add Quarters to DateTime formats

### ISSUE THREAD

**Issue by jods4** [dotnet/corefx#10994]
I suggest that you add support for quarters when formatting a `DateTime`.
`q` would display the date quarter (1..4).

Example: 
`new DateTime(2016, 8, 19).ToString("Qq yyyy") == "Q3 2016"`

This is quite common in financial applications and the lack of it can sometimes require convoluted workarounds.

**Comment by jasonwilliams200OK** [comment]
Mixing the conventions from ICU http://userguide.icu-project.org/formatparse/datetime and .NET https://msdn.microsoft.com/en-us/library/8kb3ddd4.aspx, for quarter 3:

Directive | Result
---|---
| q | 3 |
| qq | 03 |
| qqq | 3rd quarter |
| qqqq | Third quarter |

Not sure if the inverted casing `Q` is required for anything.

---

Note that ICU user guide has a bug, where it says:

> "Stand Alone" values refer to those designed to stand on their own, as opposed to being with other formatted values. "2nd quarter" would use the stand alone format (QQQQ), whereas "2nd quarter 2007" would use the regular format (qqqq yyyy).

Correction would be:

> "Stand Alone" values refer to those designed to stand on their own, as opposed to being with other formatted values. "2nd quarter" would use the stand alone format (<del>QQQQ</del> qqqq), whereas "2nd quarter 2007" would use the regular format (<del>qqqq</del> QQQQ yyyy).

Because the table below is correct. (not sure where to report that)

**Comment by MgSam** [comment]
The problem is that quarters are not necessarily standardized very well. Different companies use different quarter ends for their financial reporting, for example.

**Comment by jods4** [comment]
@MgSam Yeah fiscal year quarters depend on when your fiscal year begins and it could be any month.

Calendar year quarters are common and useful nonetheless.

**Comment by danmosemsft** [comment]
@tarekgh is this ready to be marked api-readyfor review?

**Comment by tarekgh** [comment]
@danmosemsft the issue is labeled with api-needs-work because we need to figure out how we can format dates with non English cultures. like for example if we format with Japanese culture how we can get the localized terms like "Q", "3rd quarter" and "Third Quarter". 

Today we get all needed information from the underlying OS and currently Windows doesn't support this so we don't have a good way to support it on Windows.

**Comment by jods4** [comment]
Glad to see some work on this! If Windows can't provide localizations for _Q_ and _Quarter_, having just `q` and `qq` would already be a great step forward in many scenarios.
In particular, many apps (notably LOB) are meant for a single audience and not globalized. In that context one could use a string format like `Qq yyyy` and get `Q4 2016`. That's immensely helpful in contexts where you can only use a string format to control the output (such as many reporting and charting controls).

Alignment with other platforms is a good point. Other than that, I note that reserving `qq` for `03` is a bit useless as a quarter never has two digits (1-4).

**Comment by tarekgh** [comment]
@jods4 there is another aspect here if we support such format which is the parsing. if we parse "3 2016" this can mean something else from what we have formatted especially if we use q, qq in more complicated pattern (e.g. "MMMM/yyyy mm:ss q"). 

if this needed urgently we can provide a sample helper method to do that till we figure out the complete story.

**Comment by jods4** [comment]
@tarekgh this is not needed urgently, in fact it's long past due. ;)
I have built some helpers to format quarters the way I want, that's not hard.

But as I pointed out the real pain point here is that you sometimes interact with libraries where a _format string_ is your only option (e.g. charting controls) and that's where it would really help to get framework support.

**Comment by AlexGhiondea** [comment]
We need a formal API proposal for this. Anyone interested?

**Comment by tarekgh** [comment]
> We need a formal API proposal for this. Anyone interested?

No, the change is to support "q" when have it in the datetime patterns. so theoretically there is no API change
